File: ocr_softmax_bench/softmax_bench.py
#!/usr/bin/env python3
"""CPU benchmark: OCR softmax baseline (2x exp) vs optimized (1x exp)."""

import math
import random
import time

SEQLEN = 128
CHAR_SIZE = 9973
WARMUP = 3
RUNS = 15


def softmax_baseline(logits, probs, seqlen, char_size):
    for j in range(seqlen):
        offset = j * char_size
        row = logits[offset : offset + char_size]
        max_logit = row[0]
        for k in range(1, char_size):
            v = row[k]
            if v > max_logit:
                max_logit = v
        sum_exp = 0.0
        for k in range(char_size):
            sum_exp += math.exp(row[k] - max_logit)
        for k in range(char_size):
            probs[offset + k] = math.exp(row[k] - max_logit) / sum_exp


def softmax_optimized(logits, probs, seqlen, char_size):
    for j in range(seqlen):
        offset = j * char_size
        row = logits[offset : offset + char_size]
        max_logit = row[0]
        for k in range(1, char_size):
            if row[k] > max_logit:
                max_logit = row[k]
        sum_exp = 0.0
        for k in range(char_size):
            e = math.exp(row[k] - max_logit)
            probs[offset + k] = e
            sum_exp += e
        inv = 1.0 / sum_exp
        for k in range(char_size):
            probs[offset + k] *= inv


def bench(fn, logits, probs):
    for _ in range(WARMUP):
        fn(logits, probs, SEQLEN, CHAR_SIZE)
    t0 = time.perf_counter()
    for _ in range(RUNS):
        fn(logits, probs, SEQLEN, CHAR_SIZE)
    return (time.perf_counter() - t0) * 1000.0 / RUNS


def main():
    n = SEQLEN * CHAR_SIZE
    random.seed(42)
    logits = [random.uniform(-10.0, 10.0) for _ in range(n)]
    probs_a = [0.0] * n
    probs_b = [0.0] * n

    softmax_baseline(logits, probs_a, SEQLEN, CHAR_SIZE)
    softmax_optimized(logits, probs_b, SEQLEN, CHAR_SIZE)
    diff = max(abs(a - b) for a, b in zip(probs_a, probs_b))

    base_ms = bench(softmax_baseline, logits, probs_a)
    opt_ms = bench(softmax_optimized, logits, probs_b)
    exp_base = 2 * SEQLEN * CHAR_SIZE
    exp_opt = SEQLEN * CHAR_SIZE

    print("OCR softmax CPU benchmark (Python math.exp, same logic as OpenCL kernel)")
    print(f"  seqlen={SEQLEN}  char_size={CHAR_SIZE}  elements={n}")
    print(f"  exp calls: baseline={exp_base}  optimized={exp_opt}")
    print(f"  correctness max|diff|={diff:.3e} {'PASS' if diff < 1e-5 else 'FAIL'}")
    print(f"  baseline  avg={base_ms:.2f} ms")
    print(f"  optimized avg={opt_ms:.2f} ms")
    print(f"  speedup={base_ms / opt_ms:.2f}x")
    print()
    print("Note: Python is slower than native/GPU; speedup ratio is what matters here.")


if __name__ == "__main__":
    main()

File: ocr_softmax_bench/softmax_bench.js
// OCR softmax benchmark: baseline (2x exp) vs optimized (1x exp)
// seqlen=128, char_size=9973

const SEQLEN = 128;
const CHAR_SIZE = 9973;
const WARMUP = 5;
const RUNS = 30;

function softmaxBaseline(logits, probs, seqlen, charSize) {
  for (let j = 0; j < seqlen; j++) {
    const offset = j * charSize;
    let maxLogit = logits[offset];
    for (let k = 1; k < charSize; k++) {
      const v = logits[offset + k];
      if (v > maxLogit) maxLogit = v;
    }
    let sumExp = 0;
    for (let k = 0; k < charSize; k++) {
      sumExp += Math.exp(logits[offset + k] - maxLogit);
    }
    for (let k = 0; k < charSize; k++) {
      probs[offset + k] = Math.exp(logits[offset + k] - maxLogit) / sumExp;
    }
  }
}

function softmaxOptimized(logits, probs, seqlen, charSize) {
  for (let j = 0; j < seqlen; j++) {
    const offset = j * charSize;
    let maxLogit = logits[offset];
    for (let k = 1; k < charSize; k++) {
      if (logits[offset + k] > maxLogit) maxLogit = logits[offset + k];
    }
    let sumExp = 0;
    for (let k = 0; k < charSize; k++) {
      const e = Math.exp(logits[offset + k] - maxLogit);
      probs[offset + k] = e;
      sumExp += e;
    }
    const inv = 1 / sumExp;
    for (let k = 0; k < charSize; k++) {
      probs[offset + k] *= inv;
    }
  }
}

function bench(fn, logits, probs) {
  for (let i = 0; i < WARMUP; i++) fn(logits, probs, SEQLEN, CHAR_SIZE);
  const t0 = performance.now();
  for (let i = 0; i < RUNS; i++) fn(logits, probs, SEQLEN, CHAR_SIZE);
  return (performance.now() - t0) / RUNS;
}

const n = SEQLEN * CHAR_SIZE;
const logits = new Float64Array(n);
const probsA = new Float64Array(n);
const probsB = new Float64Array(n);

for (let i = 0; i < n; i++) logits[i] = (Math.random() * 2 - 1) * 10;

softmaxBaseline(logits, probsA, SEQLEN, CHAR_SIZE);
softmaxOptimized(logits, probsB, SEQLEN, CHAR_SIZE);

let diff = 0;
for (let i = 0; i < n; i++) diff = Math.max(diff, Math.abs(probsA[i] - probsB[i]));

const baseMs = bench(softmaxBaseline, logits, probsA);
const optMs = bench(softmaxOptimized, logits, probsB);
const expBase = 2 * SEQLEN * CHAR_SIZE;
const expOpt = SEQLEN * CHAR_SIZE;

console.log("OCR softmax benchmark (Node.js, same logic as your OpenCL kernel)");
console.log(`  host: ${process.platform} ${process.arch}  node=${process.version}`);
console.log(`  seqlen=${SEQLEN}  char_size=${CHAR_SIZE}  elements=${n}`);
console.log(`  exp calls: baseline=${expBase}  optimized=${expOpt}`);
console.log(`  correctness max|diff|=${diff.toExponential(3)} ${diff < 1e-10 ? "PASS" : "FAIL"}`);
console.log(`  baseline  avg=${baseMs.toFixed(2)} ms`);
console.log(`  optimized avg=${optMs.toFixed(2)} ms`);
console.log(`  speedup=${(baseMs / optMs).toFixed(2)}x`);
console.log(`  est. GPU kernel savings: ~${((1 - expOpt / expBase) * 100).toFixed(0)}% less exp() work`);
